examples/example-server-lib/splash_session.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright © 2018 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Alan Griffiths <alan@octopull.co.uk>
+ */
+
+#ifndef MIR_SPLASH_SESSION_H
+#define MIR_SPLASH_SESSION_H
+
+#include <mir_toolkit/client_types.h>
+
+#include <memory>
+
+namespace mir { class Server; namespace scene { class Session; }}
+
+class SplashSession
+{
+public:
+    struct Self;
+    SplashSession(std::shared_ptr<Self> const self);
+
+    auto session() const -> std::shared_ptr<mir::scene::Session>;
+
+private:
+    std::shared_ptr<Self> const self;
+};
+
+class SplashSession::Self
```

Comment:
are we using PImpl here, especially when the implementation is in the same file?

Reply:
The reason to have separate `SplashSession` and `SplashSession::Self` types here is ABI stability, yes?

You can later add a `SplashSession::SelfV2` struct with more virtual functions and `SplashSession` can `dynamic_cast` its way to using them.